### PR TITLE
HIP-584: Copy token update logic

### LINKED ISSUE #6393: HIP-584: Copy `TokenUpdateLogic`
- Copy `TokenUpdateLogic` in `com.hedera.services`.
- Integrate `TokenUpdateLogic` with the Store interface.
- Add javadoc describing what has changed and how we integrated it in the `mirror-node`’s code base
- Create integration tests in `ContractCallServicePrecompileTest` in `hedera-mirror-web3` and validate that the return gas estimates in the desired range with offset between 5% and 20% more than the actual gas used

### PR DESCRIPTION
**Pull request by dikel**
**Description**:
This PR adds:
- Added `maxMemoUtf8Bytes` `maxTokenSymbolUtf8Bytes` `maxTokenNameUtf8Bytes` `minAutoRenewDuration` `maxAutoRenewDuration` `allowTreasuryToOwnNfts` EVM properties
- Copied `ImmutableKeyUtils` from `hedera-services`
- Copied `TokenUpdateLogic` from `hedera-services`
- Copied missing methods to `HederaTokenStore` from `hedera-services`
- Copied `TokenUpdateValidator` from `hedera-services`
- Copied missing methods to `OptionValidator` from `hedera-services`
- Copied `PureValidation` from `hedera-services`
- Copied missing methods to `TokenListChecks` from `hedera-services`
- Added tests for `ImmutableKeyUtils`, `TokenUpdateLogic`, `HederaTokenStore`, `ContextOptionValidator`, `TokenListChecks`

**Related issue(s)**:

Fixes #6393 

**Notes for reviewer**:

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/properties/MirrorNodeEvmProperties.java [hunk 1]
```diff
@@ -111,6 +115,24 @@ public class MirrorNodeEvmProperties implements EvmProperties {
     @Getter
     private boolean limitTokenAssociations = false;
 
+    @Getter
+    private int maxMemoUtf8Bytes = 100;
+
+    @Getter
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/properties/MirrorNodeEvmProperties.java:121]: Should probably add `@Min(1)` to all of these.

**Review by steven-sheehy — DISMISSED** [review]
LGTM. Should also have someone else from LC review

File: hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/properties/MirrorNodeEvmProperties.java [hunk 2]
```diff
@@ -111,6 +115,29 @@ public class MirrorNodeEvmProperties implements EvmProperties {
...
+    @Min(1)
+    private long minAutoRenewDuration = 1000L;
+
+    @Getter
+    @Min(1)
+    private long maxAutoRenewDuration = 10000L;
+
+    @Getter
+    private boolean allowTreasuryToOwnNfts = true;
```
Comment (IvanKavaldzhiev) [inline hedera-mirror-web3/src/main/java/com/hedera/mirror/web3/evm/properties/MirrorNodeEvmProperties.java:139]: Organise alphabetically all properties

**Review by IvanKavaldzhiev — APPROVED** [review]
LGTM - needs to address merge conflicts

**Review by agadzhalov — DISMISSED** [review]
LGTM

File: hedera-mirror-web3/src/test/java/com/hedera/services/store/contracts/precompile/TokenUpdateLogicTest.java [hunk 3]
```diff
@@ -0,0 +1,666 @@
...
+class TokenUpdateLogicTest {
+    private static final long CONSENSUS_TIME = 1_234_567L;
+    private static final TokenID fungible = IdUtils.asToken("0.0.888");
+    private static final TokenID nonFungible = IdUtils.asToken("0.0.889");
+    private static final AccountID accountId = IdUtils.asAccount("0.0.3");
+    private static final AccountID treasury = IdUtils.asAccount("0.0.4");
+    private static final NftId nftId =
+            new NftId(nonFungible.getShardNum(), nonFungible.getRealmNum(), nonFungible.getTokenNum(), -1);
+    private static final Timestamp EXPIRY = Timestamp.getDefaultInstance();
+    private static final EntityId treasuryId = EntityId.of(treasury);
```
Comment (kselveliev) [inline hedera-mirror-web3/src/test/java/com/hedera/services/store/contracts/precompile/TokenUpdateLogicTest.java:74]: I think we can delete this treasuryId, I do not see it being used in this test class

**Review by kselveliev — APPROVED** [review]

**Pull request merged**